nautobot_secrets_providers/connectors/hashicorp.py

```diff
@@ -0,0 +1,135 @@
+        # so we use a parameter to specify the path to the ca_cert, if not provided we use the default of None
+        ca_cert = cls.vault_settings.get("ca_cert", None)
+
+        # Get the client and attempt to retrieve the secret.
+        try:
+            if cls.auth_method == "token":
+                cls.client = hvac.Client(
+                    url=cls.vault_settings["url"], token=cls.vault_settings["token"], verify=ca_cert
+                )
+            else:
+                cls.client = hvac.Client(url=cls.vault_settings["url"], verify=ca_cert)
+                if cls.auth_method == "approle":
+                    cls.client.auth.approle.login(
+                        role_id=cls.vault_settings["role_id"],
+                        secret_id=cls.vault_settings["secret_id"],
+                        **login_kwargs,
+                    )
+                elif cls.auth_method == "kubernetes":
+                    with open(cls.k8s_token_path, "r", encoding="utf-8") as token_file:
+                        jwt = token_file.read()
+                    cls.client.auth.kubernetes.login(role=cls.vault_settings["role_name"], jwt=jwt, **login_kwargs)
+                elif cls.auth_method == "aws":
+                    session = boto3.Session()
+                    aws_creds = session.get_credentials()
+                    cls.client.auth.aws.iam_login(
+                        access_key=aws_creds.access_key,
+                        secret_key=aws_creds.secret_key,
+                        session_token=aws_creds.token,
+                        region=session.region_name,
+                        role=cls.vault_settings.get("role_name", None),
+                        **login_kwargs,
+                    )
+        except hvac.exceptions.InvalidRequest as err:
+            raise ConnectorError(
+                cls, f"HashiCorp Vault Login failed (auth_method: {cls.auth_method}). Error: {err}"
+            ) from err
+        except hvac.exceptions.Forbidden as err:
+            raise ConnectorError(
+                cls, f"HashiCorp Vault Access Denied (auth_method: {cls.auth_method}). Error: {err}"
+            ) from err
+
+        return cls.client
+
+    @classmethod
+    def get_kv_value(cls, secret_key, default_value=None, **kwargs):
+        """Get a kv key and return the value stored in secret_key."""
+        client = cls.get_client()
+        try:
+            response = client.secrets.kv.read_secret(**kwargs)
+        except hvac.exceptions.InvalidPath as err:
+            raise ConnectorError(cls, str(err)) from err
+
+        # Retrieve the value using the key or complain loudly.
+        try:
+            return response["data"]["data"][secret_key]
+        except KeyError as err:
+            if default_value:
+                return default_value
+            else:
+                msg = f"The secret value could not be retrieved using key {err}"
+                raise SecretValueNotFoundError(cls, msg) from err
```

Comment:
This entire module is beautiful. My only gripe is that it's duplicating patterns and/or code from the core provider source. I like where you're going with this, but is there any way we can marry these interfaces?

Reply:
I like the idea you're getting at in other words, but I don't like that it seemingly adds a lot of apparent boiler plate code, potentially complicating the introduction and maintenance of other providers.

Reply:
I'll probably re-work this entire PR but I've started the conversation to get this into Core here https://github.com/nautobot/nautobot/issues/2469